cpp/src/arrow/compute/exec/asof_join_node.cc

```diff
@@ -459,17 +642,58 @@ class CompositeReferenceTable {
     if (!_ptr2ref.count((uintptr_t)ref.get())) _ptr2ref[(uintptr_t)ref.get()] = ref;
   }
 
-  template <class Builder, class PrimitiveType>
-  Result<std::shared_ptr<Array>> MaterializePrimitiveColumn(MemoryPool* memory_pool,
-                                                            size_t i_table,
-                                                            col_index_t i_col) {
-    Builder builder(memory_pool);
+  // this should really be a method on ArrayData
+  static bool IsNull(const std::shared_ptr<ArrayData>& source, row_index_t row) {
+    return ((source->buffers[0] != NULLPTR)
+                ? !bit_util::GetBit(source->buffers[0]->data(), row + source->offset)
+                : source->null_count.load() == source->length);
+  }
+
+  template <typename T>
+  using is_fixed_width_type = std::is_base_of<FixedWidthType, T>;
+
+  template <typename T, typename R = void>
+  using enable_if_fixed_width_type = enable_if_t<is_fixed_width_type<T>::value, R>;
```

Comment:
Should you add these to `src/arrow/type_traits.h`?

Reply:
With your approval.

Reply:
I'll move these.